### PR TITLE
CBL-4096: Replicator logs collection info

### PR DESCRIPTION
**Pull request by callumbirks**
Make log functions inline instead of macro.
Add log overrides for Worker that insert collection info.
Add new log functions for Replicator that take collection index as a parameter.
Add format() override that takes format string as a std::string.

I've run the tests and reviewed the logs, all log types that were in the ticket now have collection info. I've also performance tested using unix `time` command, there seems to be no performance impact of changing the log functions from macro to virtual inline, and using overrides of the virtual functions.

### CODE REVIEW

**Comment by callumbirks** [comment]
The `pragma diagnostic` lines are to silence the "format string is not a literal string" warnings, which Xcode treats as an error.

**Comment by callumbirks** [comment]
I don't love the solution of having to silence a warning, but I think this is all we can do for now, without completely revamping the Logging. (Which is something I'd like to do)

File: Replicator/Replicator.hh [hunk 1]
```diff
@@ -125,6 +125,41 @@ namespace litecore { namespace repl {
...
+        // pass the collectionIndex manually for log calls
+        inline void cLogInfo(CollectionIndex idx, const char * fmt, ...) const {
+            const std::string fmt_ = formatWithCollection(fmt, idx);
+            va_list args;
+            va_start(args, fmt);
+            _logAt(Info, fmt_.c_str(), args);
+            va_end(args);
+        }
+        inline void cLogVerbose(CollectionIndex idx, const char * fmt, ...) const {
+            const std::string fmt_ = formatWithCollection(fmt, idx);
```
Comment (jianminzhao) [inline Replicator/Replicator.hh:146]: Nice: absorbing the index into a new format string!

File: Replicator/ReplicatorTypes.hh [hunk 2]
```diff
@@ -132,4 +132,6 @@ namespace litecore { namespace repl {
     using CollectionIndex = unsigned;
     constexpr const CollectionIndex kNotCollectionIndex
         = std::numeric_limits<CollectionIndex>::max();
+    // The log format string for logging a collection index
+    constexpr std::string_view kCollectionLogFormat = "{Coll#%i}";
```
Comment (jianminzhao) [inline Replicator/ReplicatorTypes.hh:136]: This probably belongs to Worker.hh.

**Review by jianminzhao — APPROVED** [review]

File: LiteCore/Support/StringUtil.cc [hunk 3]
```diff
@@ -50,6 +50,16 @@ namespace litecore {
         return result;
     }
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-nonliteral"
+    std::string format(const std::string fmt, ...) {
```
[the remark this answers is not in the document]
Comment (callumbirks) [reply LiteCore/Support/StringUtil.cc:55]: Unfortunately not, due to the variadic arguments, the prior function will be unable to call this one. Once a function has received the va_args, they must be expanded. Similar issue to the one we ran into with expanding the va_args twice from the log function.

File: LiteCore/Support/Logging.hh [hunk 4]
```diff
@@ -202,6 +202,10 @@ extern LogDomain DBLog, QueryLog, SyncLog, &ActorLog;
 #endif
 #endif
 
+#define _logAt(LEVEL, FMT, VA_ARGS) do { \
+    if (_usuallyFalse(this->willLog(litecore::LogLevel::LEVEL))) \
```
[the remark this answers is not in the document]
Comment (callumbirks) [reply LiteCore/Support/Logging.hh:206]: I've decided to change this func to inline, because I realised it doesn't need to be a macro anymore.

**Pull request merged**